docs/proposals/volumes.md

```diff
@@ -0,0 +1,428 @@
+
+We will not attempt to address the actual ownership and permissions concerns of new shared storage
+in this proposal.
+
+#### Plugin API requirements
+
+From the above, we know that some volume plugins will 'want' ownership management from the Kubelet
+and others will not.  Plugins should be able to opt in to ownership management from the Kubelet.  To
+facilitate this, there should be a method added to the `volume.Plugin` interface that the Kubelet
+uses to determine whether to perform ownership management for a volume.
+
+### Determining correct ownership of a volume
+
+Using the approach of a pod-level supplemental group to own volumes solves the problem in any of the
+cases of UID/GID combinations within a pod. Since this is the simplest approach that handles all
+use-cases, our solution will be made in terms of it.
+
+Eventually, Kubernetes should allocate a unique group for each pod so that a pod's volumes are
+usable by that pod's containers, but not by containers of another pod.  The supplemental group used
+to share volumes must be unique in a multitenant cluster.  If uniqueness is enforced at the host
+level, pods from one host may be able to use shared filesystems meant for pods on another host.
+
+An admission controller could handle allocating groups for each pod and setting the group in the
+pod's security context.
+
+#### A note on the root group
+
+Today, by default, all docker containers are run in the root group (GID 0).  This is relied on by
+image authors that make images to run with a range of UIDs: they set the group ownership for
+important paths to be the root group, so that containers running as GID 0 *and* an arbitrary UID
+can read and write to those paths normally.
+
+It is important to note that the changes proposed here will not affect the primary GID of
+containers in pods.  Setting the `pod.Spec.SecurityContext.SupplementalGroup` field will not
+override the primary GID and should be safe to use in images that expect GID 0.
+
+### Setting ownership and permissions on volumes
+
+For `EmptyDir`-based volumes and unshared storage, `chown` and `chmod` on the node are sufficient to
+set ownershp and permissions.  Shared storage is different because:
+
+1.  Shared storage may not live on the node a pod that uses it runs on
+2.  Shared storage may be externally managed
+
+## Proposed design:
+
+Our design should minimize code for handling ownership required in the Kubelet and volume plugins.
+
+### API changes
+
+We should not interfere with images that need to run as a particular UID or primary GID.  A pod
+level supplemental group allows us to express a group that all containers in a pod run as in a way
+that is orthogonal to the primary UID and GID of each container process.
+
+```go
+package api
+
+type PodSecurityContext struct {
+    // SupplementalGroup is a supplemental group that all containers in a pod run under
+    SupplementalGroup *int64 `json:"supplementalGroup"`
```

Comment:
Paul and I spoke today, expecting new notes so not responding to teh last
24 hours of email, sorry :)

On Tue, Sep 22, 2015 at 1:46 PM, Paul Morie <notifications@github.com>
wrote:

> In docs/proposals/volumes.md
> <https://github.com/kubernetes/kubernetes/pull/12944#discussion_r40141598>
> :
>
> > +## Proposed design:
> > +
> > +Our design should minimize code for handling ownership required in the Kubelet and volume plugins.
> > +
> > +### API changes
> > +
> > +We should not interfere with images that need to run as a particular UID or primary GID.  A pod
> > +level supplemental group allows us to express a group that all containers in a pod run as in a way
> > +that is orthogonal to the primary UID and GID of each container process.
> > +
> > +```go
> > +package api
> > +
> > +type PodSecurityContext struct {
> > +    // SupplementalGroup is a supplemental group that all containers in a pod run under
> > +    SupplementalGroup *int64 `json:"supplementalGroup"`
>
> @smarterclayton <https://github.com/smarterclayton>
>
> You can have multiple supplemental groups.
>
> To be clear I am not trying to solve supporting arbitrary supplemental
> groups here. I want a special supplemental group that all containers run
> with. I think we should do arbitrary supplemental groups and primary GID
> control in another issue.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/12944/files#r40141598>.
>